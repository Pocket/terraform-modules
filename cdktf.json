{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
