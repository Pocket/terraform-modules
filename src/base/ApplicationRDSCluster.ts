import { Resource } from 'cdktf';
import {
  DataAwsVpc,
  DbSubnetGroup,
  RdsCluster,
  RdsClusterConfig,
  SecretsmanagerSecret,
  SecretsmanagerSecretVersion,
  SecurityGroup,
} from '@cdktf/provider-aws';
import { Construct } from 'constructs';
import crypto from 'crypto';

//Override the default rds config but remove the items that we set ourselves.
export type ApplicationRDSClusterConfig = Omit<
  RdsClusterConfig,
  | 'clusterIdentifierPrefix'
  | 'vpcSecurityGroupIds'
  | 'dbSubnetGroupName'
  | 'copyTagsToSnapshot'
  | 'tags'
  | 'lifecycle'
> & {
  // masterUsername is not a required field in the RdsClusterConfig type but is required for create an RDS cluster
  masterUsername: string;
  masterPassword?: string;
  engine?: 'aurora' | 'aurora-mysql' | 'aurora-postgresql';
};

export interface ApplicationRDSClusterProps {
  prefix: string;
  vpcId: string;
  subnetIds: string[];
  rdsConfig: ApplicationRDSClusterConfig;
  tags?: { [key: string]: string };
}

/**
 * Generates an RDS cluster
 *
 * The database will be initialized with a random password.
 *
 * If the database is Aurora or MySQL, a SecretsManager secret will be created with a rotation lambda
 * that you can invoke in the AWS console after creation to rotate the password
 */
export class ApplicationRDSCluster extends Resource {
  public readonly rds: RdsCluster;
  public readonly secretARN?: string;

  constructor(
    scope: Construct,
    name: string,
    config: ApplicationRDSClusterProps
  ) {
    super(scope, name);

    const vpc = new DataAwsVpc(this, `vpc`, {
      filter: [
        {
          name: 'vpc-id',
          values: [config.vpcId],
        },
      ],
    });

    // Set the default port for mysql/postgresql based on the engine value for RDS
    const rdsPort = config.rdsConfig.engine?.includes('postgresql')
      ? 5432
      : 3306;

    const securityGroup = new SecurityGroup(this, 'rds_security_group', {
      namePrefix: config.prefix,
      description: 'Managed by Terraform',
      vpcId: vpc.id,
      ingress: [
        {
          fromPort: rdsPort,
          toPort: rdsPort,
          protocol: 'tcp',
          cidrBlocks: [vpc.cidrBlock],
          // the following are included due to a bug
          // https://github.com/hashicorp/terraform-cdk/issues/223
          description: null,
          ipv6CidrBlocks: null,
          prefixListIds: null,
          securityGroups: null,
        },
      ],
      egress: [
        {
          fromPort: 0,
          protocol: '-1',
          toPort: 0,
          cidrBlocks: ['0.0.0.0/0'],
          description: 'required',
          ipv6CidrBlocks: [],
          prefixListIds: [],
          securityGroups: [],
        },
      ],
    });

    const subnetGroup = new DbSubnetGroup(this, 'rds_subnet_group', {
      namePrefix: config.prefix.toLowerCase(),
      subnetIds: config.subnetIds,
    });

    this.rds = new RdsCluster(this, 'rds_cluster', {
      ...config.rdsConfig,
      clusterIdentifierPrefix: config.prefix.toLowerCase(),
      tags: config.tags,
      copyTagsToSnapshot: true, //Why would we ever want this to false??
      masterPassword:
        config.rdsConfig.masterPassword ??
        crypto.randomBytes(8).toString('hex'),
      vpcSecurityGroupIds: [securityGroup.id],
      dbSubnetGroupName: subnetGroup.name,
      lifecycle: {
        ignoreChanges: ['master_username', 'master_password'],
      },
    });

    // Create secrets manager resource for the RDS
    // This value should be changed after initial creation
    const { secretARN } = ApplicationRDSCluster.createRdsSecret(
      this,
      this.rds,
      rdsPort,
      config.prefix,
      config.tags,
      config.rdsConfig.engine
    );

    this.secretARN = secretARN;
  }

  /**
   * Create an RDS secret
   *
   * @param scope
   * @param rds
   * @param rdsPort
   * @param prefix
   * @param tags
   * @private
   */
  private static createRdsSecret(
    scope: Construct,
    rds: RdsCluster,
    rdsPort: number,
    prefix: string,
    tags?: { [key: string]: string },
    engine?: ApplicationRDSClusterConfig['engine']
  ): { secretARN: string } {
    //Create the secret
    const secret = new SecretsmanagerSecret(scope, `rds_secret`, {
      description: `Secret For ${rds.clusterIdentifier}`,
      name: `${prefix}/${rds.clusterIdentifier}`,
      //We dont auto rotate, because our apps dont have triggers to refresh yet.
      //This is mainly so we can rotate after we create the RDS.
      dependsOn: [rds],
    });

    const secretValues: {
      engine: string;
      host: string;
      username: string;
      password: string;
      dbname: string;
      port: number;
      database_url?: string;
    } = {
      engine: rds.engine,
      host: rds.endpoint,
      username: rds.masterUsername,
      password: rds.masterPassword,
      dbname: rds.databaseName,
      port: rdsPort,
    };

    // Add a database URL to a MySQL-compatible Aurora instance
    if (engine && engine === 'aurora-mysql') {
      secretValues.database_url = `mysql://${rds.masterUsername}:${rds.masterPassword}@${rds.endpoint}:${rdsPort}/${rds.databaseName}`;
    }

    //Create the initial secret version
    new SecretsmanagerSecretVersion(scope, `rds_secret_version`, {
      secretId: secret.id,
      secretString: JSON.stringify(secretValues),
      dependsOn: [secret],
    });

    return { secretARN: secret.arn };
  }
}
