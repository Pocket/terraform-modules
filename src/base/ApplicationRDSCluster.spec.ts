import { Testing } from 'cdktf';
import { ApplicationRDSCluster } from './ApplicationRDSCluster';

describe('ApplicationRDSCluster', () => {
  it('renders a RDS cluster', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRDSCluster(stack, 'testRDSCluster', {
        prefix: 'bowling-',
        vpcId: 'rug',
        subnetIds: ['0', '1'],
        rdsConfig: {
          masterUsername: 'walter',
          masterPassword: 'bowling',
          databaseName: 'walter',
          engine: 'aurora-mysql',
        },
        tags: {
          whodis: 'walter',
        },
      });
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders a RDS cluster with a database URL', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRDSCluster(stack, 'testRDSCluster', {
        prefix: 'bowling-',
        vpcId: 'rug',
        subnetIds: ['0', '1'],
        rdsConfig: {
          masterUsername: 'walter',
          masterPassword: 'bowling',
          databaseName: 'walter',
          engine: 'aurora-mysql',
        },
        tags: {
          whodis: 'walter',
        },
      });
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders a RDS cluster without a name', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRDSCluster(stack, 'testRDSCluster', {
        prefix: 'bowling-',
        vpcId: 'rug',
        subnetIds: ['0', '1'],
        useName: false,
        rdsConfig: {
          masterUsername: 'walter',
          masterPassword: 'bowling',
          databaseName: 'walter',
          engine: 'aurora-mysql',
        },
        tags: {
          whodis: 'walter',
        },
      });
    });
    expect(synthed).toMatchSnapshot();
  });
});
