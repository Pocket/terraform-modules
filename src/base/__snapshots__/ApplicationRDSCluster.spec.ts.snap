// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a RDS cluster 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"data\\": {
    \\"aws_vpc\\": {
      \\"testRDSCluster_vpc_F47EEEFE\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"rug\\"
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/vpc\\",
            \\"uniqueId\\": \\"testRDSCluster_vpc_F47EEEFE\\"
          }
        }
      }
    }
  },
  \\"resource\\": {
    \\"aws_security_group\\": {
      \\"testRDSCluster_rds_security_group_4A9D257E\\": {
        \\"description\\": \\"Managed by Terraform\\",
        \\"egress\\": [
          {
            \\"cidr_blocks\\": [
              \\"0.0.0.0/0\\"
            ],
            \\"description\\": \\"required\\",
            \\"from_port\\": 0,
            \\"ipv6_cidr_blocks\\": [],
            \\"prefix_list_ids\\": [],
            \\"protocol\\": \\"-1\\",
            \\"security_groups\\": [],
            \\"self\\": null,
            \\"to_port\\": 0
          }
        ],
        \\"ingress\\": [
          {
            \\"cidr_blocks\\": [
              \\"\${data.aws_vpc.testRDSCluster_vpc_F47EEEFE.cidr_block}\\"
            ],
            \\"description\\": null,
            \\"from_port\\": 3306,
            \\"ipv6_cidr_blocks\\": null,
            \\"prefix_list_ids\\": null,
            \\"protocol\\": \\"tcp\\",
            \\"security_groups\\": null,
            \\"self\\": null,
            \\"to_port\\": 3306
          }
        ],
        \\"name_prefix\\": \\"bowling-\\",
        \\"vpc_id\\": \\"\${data.aws_vpc.testRDSCluster_vpc_F47EEEFE.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/rds_security_group\\",
            \\"uniqueId\\": \\"testRDSCluster_rds_security_group_4A9D257E\\"
          }
        }
      }
    },
    \\"aws_db_subnet_group\\": {
      \\"testRDSCluster_rds_subnet_group_88022457\\": {
        \\"name_prefix\\": \\"bowling-\\",
        \\"subnet_ids\\": [
          \\"0\\",
          \\"1\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/rds_subnet_group\\",
            \\"uniqueId\\": \\"testRDSCluster_rds_subnet_group_88022457\\"
          }
        }
      }
    },
    \\"aws_rds_cluster\\": {
      \\"testRDSCluster_rds_cluster_B5FD08B5\\": {
        \\"cluster_identifier_prefix\\": \\"bowling-\\",
        \\"copy_tags_to_snapshot\\": true,
        \\"database_name\\": \\"walter\\",
        \\"db_subnet_group_name\\": \\"\${aws_db_subnet_group.testRDSCluster_rds_subnet_group_88022457.name}\\",
        \\"master_password\\": \\"bowling\\",
        \\"master_username\\": \\"walter\\",
        \\"tags\\": {
          \\"whodis\\": \\"walter\\"
        },
        \\"vpc_security_group_ids\\": [
          \\"\${aws_security_group.testRDSCluster_rds_security_group_4A9D257E.id}\\"
        ],
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"master_username\\",
            \\"master_password\\"
          ]
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/rds_cluster\\",
            \\"uniqueId\\": \\"testRDSCluster_rds_cluster_B5FD08B5\\"
          }
        }
      }
    },
    \\"aws_secretsmanager_secret\\": {
      \\"testRDSCluster_rds_secret_A2014138\\": {
        \\"description\\": \\"Secret For \${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.cluster_identifier}\\",
        \\"name\\": \\"bowling-/\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.cluster_identifier}\\",
        \\"depends_on\\": [
          \\"aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/rds_secret\\",
            \\"uniqueId\\": \\"testRDSCluster_rds_secret_A2014138\\"
          }
        }
      }
    },
    \\"aws_secretsmanager_secret_version\\": {
      \\"testRDSCluster_rds_secret_version_55C44893\\": {
        \\"secret_id\\": \\"\${aws_secretsmanager_secret.testRDSCluster_rds_secret_A2014138.id}\\",
        \\"secret_string\\": \\"{\\\\\\"engine\\\\\\":\\\\\\"\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.engine}\\\\\\",\\\\\\"host\\\\\\":\\\\\\"\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.endpoint}\\\\\\",\\\\\\"username\\\\\\":\\\\\\"\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.master_username}\\\\\\",\\\\\\"password\\\\\\":\\\\\\"\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.master_password}\\\\\\",\\\\\\"dbname\\\\\\":\\\\\\"\${aws_rds_cluster.testRDSCluster_rds_cluster_B5FD08B5.database_name}\\\\\\",\\\\\\"port\\\\\\":3306}\\",
        \\"depends_on\\": [
          \\"aws_secretsmanager_secret.testRDSCluster_rds_secret_A2014138\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testRDSCluster/rds_secret_version\\",
            \\"uniqueId\\": \\"testRDSCluster_rds_secret_version_55C44893\\"
          }
        }
      }
    }
  }
}"
`;
